File: leetcode/Easy/go/1925. Count Square Sum Triples.go
/**
 * Решение задачи "Count Square Sum Triples" (LeetCode 1925)
 * 
 * @file count_square_sum_triples.go
 * @brief Подсчёт упорядоченных троек Пифагора в диапазоне [1, n]
 * 
 * @author Дулей Максим Игоревич
 * @see ORCID: https://orcid.org/0009-0007-7605-539X
 * @see GitHub: https://github.com/QuadDarv1ne/
 * 
 * @details
 * Алгоритм: Перебор всех пар (a, b) с проверкой через sqrt
 * Сложность: O(n²) времени, O(1) памяти
 * 
 * @approach
 * 1. Генерируем все пары (a, b) с помощью двойного цикла
 * 2. Для каждой пары вычисляем c = √(a² + b²)
 * 3. Проверяем, что c - целое число и не превышает n
 * 4. Считаем все подходящие тройки
 */

package main

import (
	"fmt"
	"math"
)

/**
 * Подсчитывает количество упорядоченных троек (a, b, c)
 * 
 * @param n Верхняя граница для значений a, b, c (1 ≤ n ≤ 250)
 * @return int Количество троек, удовлетворяющих условию a² + b² = c²
 * 
 * @example
 * result := countTriples(5)   // возвращает 2
 * result := countTriples(10)  // возвращает 4
 * 
 * @note
 * - Учитываются упорядоченные тройки: (3,4,5) и (4,3,5) - разные
 * - Используется математическая библиотека для вычисления sqrt
 * - Проверка целочисленности: c*c == a*a + b*b
 */
func countTriples(n int) int {
	count := 0

	// Перебираем все возможные значения a и b
	for a := 1; a <= n; a++ {
		for b := 1; b <= n; b++ {
			cSq := a*a + b*b
			c := int(math.Sqrt(float64(cSq)))

			// Проверяем условия Пифагоровой тройки
			if c <= n && c*c == cSq {
				count++
			}
		}
	}

	return count
}

/**
 * Оптимизированная версия с предварительным вычислением квадратов
 * 
 * @param n Верхняя граница диапазона
 * @return int Количество троек
 * 
 * @complexity
 * - Время: O(n²)
 * - Память: O(n²) для хранения карты квадратов
 */
func countTriplesOptimized(n int) int {
	// Создаём карту для быстрой проверки квадратов
	squares := make(map[int]bool)
	for i := 1; i <= n; i++ {
		squares[i*i] = true
	}

	count := 0
	maxSquare := n * n

	// Перебираем все пары квадратов
	for aSq := range squares {
		for bSq := range squares {
			sum := aSq + bSq
			if sum <= maxSquare && squares[sum] {
				count++
			}
		}
	}

	return count
}

/**
 * Версия с использованием массива вместо карты
 * Более эффективна по памяти для небольших n
 */
func countTriplesArray(n int) int {
	maxSquare := n * n
	isPerfectSquare := make([]bool, maxSquare+1)

	// Отмечаем все совершенные квадраты
	for i := 1; i <= n; i++ {
		isPerfectSquare[i*i] = true
	}

	count := 0

	// Перебираем все пары (a, b)
	for a := 1; a <= n; a++ {
		aSq := a * a
		for b := 1; b <= n; b++ {
			sum := aSq + b*b
			if sum <= maxSquare && isPerfectSquare[sum] {
				count++
			}
		}
	}

	return count
}

/**
 * Конкурентная версия для больших n
 * Использует горутины для параллельных вычислений
 */
func countTriplesConcurrent(n int) int {
	var count int32
	var wg sync.WaitGroup

	// Обрабатываем каждое a в отдельной горутине
	for a := 1; a <= n; a++ {
		wg.Add(1)
		go func(a int) {
			defer wg.Done()
			localCount := 0
			aSq := a * a

			for b := 1; b <= n; b++ {
				cSq := aSq + b*b
				c := int(math.Sqrt(float64(cSq)))
				if c <= n && c*c == cSq {
					localCount++
				}
			}

			atomic.AddInt32(&count, int32(localCount))
		}(a)
	}

	wg.Wait()
	return int(count)
}

File: leetcode/Easy/cpp/1925. Count Square Sum Triples.cpp
/**
 * @file count_square_sum_triples.cpp
 * @brief Решение задачи "Count Square Sum Triples" (LeetCode 1925)
 * 
 * @author Дулей Максим Игоревич
 * @see ORCID: https://orcid.org/0009-0007-7605-539X
 * @see GitHub: https://github.com/QuadDarv1ne/
 * 
 * @details
 * Алгоритм подсчёта упорядоченных пифагоровых троек в диапазоне [1, n]
 * Сложность: O(n²) по времени, O(1) по памяти
 * 
 * @approach
 * 1. Для каждой пары (a, b) вычисляем сумму квадратов
 * 2. Находим целую часть квадратного корня
 * 3. Проверяем, что квадрат этой части равен сумме и c ≤ n
 * 4. Учитываем упорядоченность: (a,b,c) и (b,a,c) - разные тройки
 */

#include <cmath>
#include <vector>
#include <algorithm>

class Solution {
public:
    /**
     * @brief Подсчитывает количество упорядоченных пифагоровых троек
     * 
     * @param n Верхняя граница диапазона (1 ≤ n ≤ 250)
     * @return int Количество троек, удовлетворяющих a² + b² = c²
     * 
     * @example
     *   Solution sol;
     *   int result = sol.countTriples(5);    // Возвращает 2
     *   int result = sol.countTriples(10);   // Возвращает 4
     * 
     * @note
     * - Используется проверка через целочисленный квадратный корень
     * - Тройки упорядочены: (3,4,5) и (4,3,5) считаются разными
     * - Максимальное n=250 гарантирует O(n²) решение
     */
    int countTriples(int n) {
        int count = 0;
        
        for (int a = 1; a <= n; ++a) {
            for (int b = 1; b <= n; ++b) {
                int c_sq = a * a + b * b;                // a² + b²
                int c = static_cast<int>(sqrt(c_sq));    // ⌊√(a²+b²)⌋
                
                // Проверяем, что c² = a² + b² и c ≤ n
                if (c <= n && c * c == c_sq) {
                    ++count;
                }
            }
        }
        
        return count;
    }
    
    /**
     * @brief Оптимизированная версия с предварительным вычислением квадратов
     * 
     * @param n Верхняя граница диапазона
     * @return int Количество троек
     * 
     * @complexity
     * - Время: O(n²)
     * - Память: O(n) для хранения массива квадратов
     */
    int countTriplesOptimized(int n) {
        int count = 0;
        int max_sq = n * n;
        
        // Создаём массив для отметки совершенных квадратов
        std::vector<bool> is_perfect_square(max_sq + 1, false);
        for (int i = 1; i <= n; ++i) {
            is_perfect_square[i * i] = true;
        }
        
        // Перебираем все пары (a, b)
        for (int a = 1; a <= n; ++a) {
            int a_sq = a * a;
            for (int b = 1; b <= n; ++b) {
                int sum = a_sq + b * b;
                if (sum <= max_sq && is_perfect_square[sum]) {
                    ++count;
                }
            }
        }
        
        return count;
    }
};